### PR TITLE
[DONT MERGE] Trying revapi for detecting breaking changes

### PR DESCRIPTION
**Pull request by mmollaverdi**
**Not to be merged - Only for a demo at this stage**
 
Running `gradle :misk-action-scopes:check` on this change (revapi adds a task to check), I get the following error, which is a good description of the breaking change.

```
Execution failed for task ':misk-action-scopes:revapi'.
> There were Java public API/ABI breaks reported by revapi:

  java.method.numberOfParametersChanged: The number of parameters of the method have changed.

  old: method <T, A extends java.lang.annotation.Annotation> void misk.scope.ActionScopedProviderModule::bindProvider(kotlin.reflect.KClass<T>, kotlin.reflect.KClass<? extends misk.scope.ActionScopedProvider<? extends T>>, java.lang.Class<A>)
  new: method <T, A extends java.lang.annotation.Annotation> void misk.scope.ActionScopedProviderModule::bindProvider(kotlin.reflect.KClass<T>, kotlin.reflect.KClass<? extends misk.scope.ActionScopedProvider<? extends T>>, java.lang.Class<A>, java.lang.String)

  SOURCE: BREAKING, BINARY: BREAKING

  From old archive: misk-action-scopes-2023.10.02.170330-4694630.jar
  From new archive: misk-action-scopes-0.0-SNAPSHOT.jar

  If this is an acceptable break that will not harm your users, you can ignore it in future runs like so for:
...
```

Good things to note
- Unlike `binary-compatibility-validator`, this doesn't complain about new methods being added (non-breaking changes) 🎉 
- It detected that the change to the other `bindProvider` method, where a new argument is added with a default value and an accompanying `@JvmOverloads` annotation, is not a breaking change.

### CODE REVIEW

File: .palantir/revapi.yml [hunk 1]
```diff
@@ -0,0 +1,2 @@
+versionOverrides:
+  com.squareup.misk:misk-action-scopes:misk-0.18.0: "2023.10.02.170330-4694630"
```
Comment (mmollaverdi) [inline .palantir/revapi.yml:2]: Revapi uses git tags to find the most recent version to compare against. Looks like we stopped adding git tags a long time ago. This is my hacky solution to work around that for this demo.

To use revapi for real, we'd need to either start adding git tags on release, or write version overrides to this files on release (there is gradle task provided by revapi for this). Neither sound like blocking issues?

File: misk-action-scopes/build.gradle.kts [hunk 2]
```diff
@@ -6,6 +6,7 @@ plugins {
   kotlin("jvm")
   `java-library`
   id("com.vanniktech.maven.publish.base")
+  id("com.palantir.revapi")
```
Comment (mmollaverdi) [inline misk-action-scopes/build.gradle.kts:9]: Only targeting misk-action-scopes for this demo.

File: misk-action-scopes/src/main/kotlin/misk/scope/ActionScopedProviderModule.kt [hunk 3]
```diff
@@ -27,12 +28,13 @@ abstract class ActionScopedProviderModule : KAbstractModule() {
   abstract fun configureProviders()
 
   /** Binds an [ActionScoped] which only pulls from data seeded at the scope entry */
-  fun <T : Any> bindSeedData(kclass: KClass<T>) {
+  fun <T : Any> bindSeedData(kclass: KClass<T>, newThing: String?) {
```
Comment (mmollaverdi) [inline misk-action-scopes/src/main/kotlin/misk/scope/ActionScopedProviderModule.kt:31]: Breaking change picked up by revapi

File: misk-action-scopes/src/main/kotlin/misk/scope/ActionScopedProviderModule.kt [hunk 4]
```diff
@@ -27,12 +28,13 @@ abstract class ActionScopedProviderModule : KAbstractModule() {
...
     bindSeedData(Key.get(kclass.java), Key.get(actionScopedType(kclass)))
   }
 
   /** Binds an [ActionScoped] which only pulls from data seeded at the scope entry */
-  fun <T : Any> bindSeedData(type: TypeLiteral<T>) {
+  @JvmOverloads
+  fun <T : Any> bindSeedData(type: TypeLiteral<T>, newThing: String? = null) {
```
Comment (mmollaverdi) [inline misk-action-scopes/src/main/kotlin/misk/scope/ActionScopedProviderModule.kt:37]: The combination of `@JvmOverloads` and default value for make this a on-breaking change, and revapi doesn't fail on this.

File: misk-action-scopes/src/main/kotlin/misk/scope/ActionScopedProviderModule.kt [hunk 5]
```diff
@@ -44,6 +46,8 @@ abstract class ActionScopedProviderModule : KAbstractModule() {
     )
   }
 
+  fun newShinyBindSeedData(type: KType) = bindSeedData(type.typeLiteral())
```
Comment (mmollaverdi) [inline misk-action-scopes/src/main/kotlin/misk/scope/ActionScopedProviderModule.kt:49]: Doesn't fail on new methods added.

File: misk-action-scopes/src/main/kotlin/misk/scope/ActionScopedProviderModule.kt [hunk 6]
```diff
@@ -164,7 +168,8 @@ abstract class ActionScopedProviderModule : KAbstractModule() {
   fun <T : Any, A : Annotation> bindProvider(
     kclass: KClass<T>,
     providerType: KClass<out ActionScopedProvider<T>>,
-    annotatedBy: Class<A>
+    annotatedBy: Class<A>,
+    newThing: String
```
Comment (mmollaverdi) [inline misk-action-scopes/src/main/kotlin/misk/scope/ActionScopedProviderModule.kt:172]: Breaking change picked up by revapi

File: .github/workflows/Linting.yml [hunk 7]
```diff
@@ -30,6 +30,20 @@ jobs:
...
+
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v3
+
+      - name: Hermit
+        run: ./bin/hermit env -r >> $GITHUB_ENV
+
+      - name: API/ABI Check
+        run: gradle revapi
```
Comment (mmollaverdi) [inline .github/workflows/Linting.yml:45]: Can't get this to run on CI. The task gets skipped for some reason.

**Comment by jvmakine** [comment]
This does look pretty amazing, I would be happy to have this added to Misk

**Pull request closed without merging**